dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -365,33 +365,62 @@ bool FramesDecoder::IsFormatSeekable() {
 }
 
 void FramesDecoder::BuildIndex() {
-  // TODO(awolant): Optimize this function for:
-  //  - CFR
-  //  - index present in the header
-
   index_ = vector<IndexEntry>();
 
   int last_keyframe = -1;
-  while (ReadRegularFrame(nullptr, false)) {
+  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
     IndexEntry entry;
-    entry.is_keyframe = av_state_->frame_->flags & AV_FRAME_FLAG_KEY;
-    entry.pts = av_state_->frame_->pts;
-    entry.is_flush_frame = false;
+    // We want to make sure that we call av_packet_unref in every iteration
+    auto packet = AVPacketScope(av_state_->packet_, av_packet_unref);
 
+    if (packet->stream_index != av_state_->stream_id_) {
+      continue;
+    }
+    if (packet->flags & AV_PKT_FLAG_KEY) {
+      if (index_->size() == 0) {
+        entry.is_keyframe = true;
+      }
+      auto codec_id = av_state_->ctx_->streams[packet->stream_index]->codecpar->codec_id;
+      if (codec_id == AV_CODEC_ID_H264) {
+        if (packet->size >= 5) {  // Ensure there's enough data to inspect
+          uint8_t nal_unit_type = packet->data[4] & 0x1F;
+          if (nal_unit_type == 5) {  // IDR frame for H.264
+            entry.is_keyframe = true;
+          }
+        }
+      } else if (codec_id == AV_CODEC_ID_HEVC) {
+        if (packet->size >= 6) {  // Ensure there's enough data to inspect
+          uint8_t nal_unit_type = (packet->data[4] >> 1) & 0x3F;
+          if (nal_unit_type >= 16 && nal_unit_type <= 21) {
+            entry.is_keyframe = true;
+          }
+        }
+      } else {
+        entry.is_keyframe = true;
+      }
+    } else {
+      entry.is_keyframe = false;
+    }
     if (entry.is_keyframe) {
       last_keyframe = index_->size();
     }
+    if (packet->pts != AV_NOPTS_VALUE) {
+      entry.pts = packet->pts;
+    } else {
+      entry.pts = packet->dts;
+    }
+    entry.is_flush_frame = false;
     entry.last_keyframe_id = last_keyframe;
     index_->push_back(entry);
   }
-  while (ReadFlushFrame(nullptr, false)) {
-    IndexEntry entry;
-    entry.is_keyframe = false;
-    entry.pts = av_state_->frame_->pts;
-    entry.is_flush_frame = true;
-    entry.last_keyframe_id = last_keyframe;
-    index_->push_back(entry);
-  }
+  av_packet_unref(av_state_->packet_);
```

Comment:
why do you need this? Isn't the AVPacketScope enough?

Reply:
To my understanding when av_read_frame returns >0, w still need to unref th packet and in this case we won't reach AVPacketScope creation.

Reply:
https://ffmpeg.org/doxygen/trunk/group__lavf__decoding.html#ga4fdb3084415a82e3810de6ee60e46a61
```
Returns
0 if OK, < 0 on error or end of file. On error, pkt will be blank (as if it came from av_packet_alloc()).
```